kong/runloop/balancer/init.lua

```diff
@@ -359,6 +359,16 @@ local function execute(balancer_data, ctx)
     balancer_data.balancer_handle = handle
 
   else
+    -- Note: balancer_data.retry_callback is only set by PDK once in access phase
```

Comment:
It seems that the `retry_callback` can be invoked only when `balancer == nil`, is it expected to work like this?

In addition, should we wrap retry_callback with pcall?

Reply:
> It seems that the retry_callback can be invoked only when balancer == nil, is it expected to work like this?

Yes, if we skip the balancer lookup by doing `kong.service.set_target`,  i.e. currently request doesn't have a upstream, then `balancer, upstream = balancers.get_balancer(balancer_data)` will return no balancer. Actually, this else branch is just made to support `kong.service.set_target`.
This matches our expectation, retry_callback will not be called if Kong's balancer is not skipped.

> In addition, should we wrap retry_callback with pcall?

yes, adding that now.

Reply:
I will address all comments in EE side of PR and port here

Reply:
fixed